src/htm/algorithms/Connections.cpp

```diff
@@ -534,15 +534,72 @@ void Connections::raisePermanencesToThreshold(
   if( increment <= 0 ) // If minPermSynPtr is already connected then ...
     return;            // Enough synapses are already connected.
 
-  // Raise the permance of all synapses in the potential pool uniformly.
-  for( const auto &syn : synapses ) //TODO vectorize: vector + const to all members
-    updateSynapsePermanence(syn, synapses_[syn].permanence + increment); //this is performance HOTSPOT
+  // Raise the permanence of all synapses in the potential pool uniformly.
+  bumpSegment(segment, increment);
+}
+
+
+void Connections::synapseCompetition(
+                    const Segment    segment,
+                    const SynapseIdx minimumSynapses,
+                    const SynapseIdx maximumSynapses)
+{
+  NTA_ASSERT( minimumSynapses <= maximumSynapses);
+
+  const auto &segData = dataForSegment( segment );
+
+  if( segData.synapses.empty())
+    return;   // No synapses to work with, no work to do.
+
+  // Sort the potential pool by permanence values, and look for the synapse with
+  // the N'th greatest permanence, where N is the desired number of connected
+  // synapses.  Then calculate how much to change the N'th synapses permance by
+  // such that it becomes a connected synapse.  After that there will be exactly
```

Comment:
> how much to change the N'th synapses permance by such that it becomes a connected synapse

Unrelated to acceptance of this PR, but: what is the biological mechanism for this? I am doing some work on at SP now and realized I "dont like" `raisePermanencesToThreshold`. The biological synapse cannot have this, and grow bit by bit (synPermActiveInc).

Reply:
@ctrl-z-9000-times ping on this PR, I'm looking forward to have it merge, when you have time to integrated the feedback :angel: 

Still thinking about this PR: 

- 1. please make it used in SP, TM. 
- 2. consider the following: 
  - instead of `bumpSegment` that puts all misperforming synapses in line (to min/max-desired)
  - have them die out.
  - if too few synapses on segment, create a random new. Or create each time one is removed/dead.